clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -160,6 +157,9 @@ public MemoryRecords build() {
      * Get the max timestamp and its offset. If the log append time is used, then the offset will
      * be either the first offset in the set if no compression is used or the last offset otherwise.
      * @return The max timestamp and its offset
+     *
+     * I don't really understand the logic of this code.
```

Comment:
Heh, that's a bit worrying. I'm assuming your question re: `MagicValue_V0` is re: timestamps? I haven't done an exhaustive search, but it seems like we're always passing in a `TimestampType` even with older message versions so that this isn't an issue. But an assertion here might help validate that and ensure we never violate that assumption.